### PR TITLE
Refactor distributed

### PR DESCRIPTION
**Pull request by stefansc1**
Use any SpiceEV strategy inside the distributed strategy.
No individual load_vehicle-function needed anymore.
Changes to balanced and greedy for performance issues (remove iterative calculation from balanced, simplify charging logic because load_vehicle function was removed).

### CODE REVIEW

File: spice_ev/strategies/balanced_market.py [hunk 1]
```diff
@@ -8,11 +8,10 @@
 class BalancedMarket(Strategy):
     """ Price oriented charging at times of low energy price. """
     def __init__(self, components, start_time, **kwargs):
-        self.PRICE_THRESHOLD = 0.001  # EUR/kWh
         self.HORIZON = 24  # maximum number of hours ahead
 
         super().__init__(components, start_time, **kwargs)
-        assert len(self.world_state.grid_connectors) == 1, "Only one grid connector supported"
+        # assert len(self.world_state.grid_connectors) == 1, "Only one grid connector supported"
```
Comment (jakobgemassmer) [inline spice_ev/strategies/balanced_market.py:14]: why is this a comment and not removed entirely?

File: spice_ev/strategies/greedy.py [hunk 2]
```diff
@@ -37,66 +37,47 @@ def step(self):
...
+            power = 0
+            bat_power_used = False
+            avg_power = 0
+
+            if get_cost(1, gc.cost) <= self.PRICE_THRESHOLD:
+                # low energy price: take max available power from GC (without stationary batteries)
+                power = clamp_power(gc_power_left, vehicle, cs)
+                # charge with power
+                avg_power = vehicle.battery.load(self.interval, power)['avg_power']
+            elif delta_soc > 0:
```
Comment (jakobgemassmer) [inline spice_ev/strategies/greedy.py:62]: why not check for `delta_soc > self.EPS` as in balanced.py?

File: spice_ev/strategies/greedy.py [hunk 3]
```diff
@@ -37,66 +37,47 @@ def step(self):
...
+                bat_power_used = True
+
+            # add load to charging station
+            charging_stations[cs_id] = gc.add_load(cs_id, avg_power)
+            cs.current_power += avg_power
+
+            if bat_power_used:
+                # check for power from stationary battery
+                avail_bat_power[gc_id] = max(avail_bat_power[gc_id] - avg_power, 0)
+
```
Comment (jakobgemassmer) [inline spice_ev/strategies/greedy.py:78]: balanced.py has these two checks. At least the second check seems useful here to:
```
            # can active charging station bear minimum load?
            assert cs.max_power >= cs.current_power - self.EPS, (
                "{} - {} over maximum load ({} > {})".format(
                    self.current_time, cs_id, cs.current_power, cs.max_power))
            # can grid connector bear load?
            assert gc.cur_max_power >= gc.get_current_load() - self.EPS, (
                "{} - {} over maximum load ({} > {})".format(
                    self.current_time, cs.parent, gc.get_current_load(), gc.cur_max_power))
```

**Review by jakobgemassmer — CHANGES_REQUESTED** [review]
Harmonized greedy and balanced and added three questions

**Comment by jakobgemassmer** [comment]
In strategy balanced the `PRICE_THRESHOLD` is removed, since it is set in `strategy.py`. Shouldn't this be removed in greedy, balanced_market and distributed as well?

File: spice_ev/strategies/balanced_market.py (hunk 1 above)
Comment (stefansc1) [reply spice_ev/strategies/balanced_market.py:14]: was just a temporary solution. See PR #187

**Review by jakobgemassmer — APPROVED** [review]

**Pull request merged**